src/de/jost_net/JVerein/server/DDLTool/Updates/Update0452.java

```diff
@@ -0,0 +1,34 @@
+/**********************************************************************
+ * This program is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation, either version 3 of the 
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without 
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See 
+ * the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program. If not, 
+ * see <http://www.gnu.org/licenses/>.
+ * 
+ **********************************************************************/
+package de.jost_net.JVerein.server.DDLTool.Updates;
+
+import de.jost_net.JVerein.server.DDLTool.AbstractDDLUpdate;
+import de.willuhn.util.ApplicationException;
+import de.willuhn.util.ProgressMonitor;
+
+import java.sql.Connection;
+
+public class Update0452 extends AbstractDDLUpdate
+{
+  public Update0452(String driver, ProgressMonitor monitor, Connection conn)
+  {
+    super(driver, monitor, conn);
+  }
+
+  @Override
+  public void run() throws ApplicationException
+  {
+    execute("update mitgliedskonto set betrag = 0 where betrag IS NULL");
```

Comment:
Wenn man die Datenbank anfasst, könnte man gleich noch einen Schritt weitergehen:

```sql
ALTER TABLE mitgliedskonto CHANGE COLUMN betrag betrag DOUBLE NOT NULL DEFAULT 0;
```

Und das `CREATE TABLE` statement auch anpassen. Oder was meint ihr?

Reply:
Ich habe das jetzt mit SET NOT NULL gemacht.
Die Kommandozeile im Vorschlag hat nicht funktioniert. Auch muss man zwischen H2DB und MySQL unterscheiden.
Es gab bereits eine Methode alterColumnNotNull die ich verwende. Sie hatte allerdings zwei Fehler. Es hat auf NULL gesetzt und nicht auf NOT NULL und dann hat auch noch das ; gefehlt.
Zumintest das NOT NULL ist jetzt gesetzt. Das mit dem DEFAULT habe ich nicht probiert wie das noch geht. Zumindest in dem gleichen Kommando ging es nicht. Aber so sollte es jetzt auch reichen.

Reply:
Bei mir ging es mit MySQL 8. Vermutlich funktioniert es nicht überall, weil `DEFAULT 0` vor `NOT NULL` stehen sollte. So unterstützt es auch die H2DB (siehe [Column definition](https://h2database.com/html/grammar.html#column_definition)).
Angeben sollten wir `DEFAULT 0` schon, da im strict mode sonst ein Fehler kommt.

Reply:
Deine Referenz ist für ein Tabellen erzeugen. Bei alter table scheint man ein SET DEFAULT 0 zu brauchen. https://www.h2database.com/html/commands.html#alter_table_alter_column

Ich kann das später mal ausprobieren.

Bei MySQL habe ich eine MariaDb bei mir zum testen.

Wo ist eigentlich das Problem wenn der Defaultwert nicht gesetzt ist? Beim Erzeugen eines Eintrags muss man halt einen Wert setzen. Oder hat das was mit Internas der DB zu tun?

Reply:
Ich habe mir die Attribute Datum und Mitglied angeschaut. Auch diese haben NOT NULL und keinen Default Wert. Dann sollte es auch beim Betrag so reichen.

Ich kann auch Sollbuchungen erzeugen wenn ein Betrag gesetzt ist. Ist keiner gesetzt kommt es zur Exception.
Das ist doch aber genau was ich haben wollte. Man sollte keine Sollbuchung ohne Betrag erzeugen können. Erstens wird es sowieso im Code geprüft, dass ein Wert eingegeben wurde und den NOT NULL habe ich so verstanden, dass zusätzlich auch die DB das nicht zulassen soll. Genau das macht sie wie erwartet.
Ein default Wert von 0 würde das dann ja wieder aushebeln.

Aus den beiden Gründen sehe darum jetzt keinen Anlass einen Default Wert hier zu setzen.

Reply:
Okay, dann hoffen wir mal, dass wir alle Exceptions abfangen und entsprechend behandeln.